extensions/workflows/requirements/cwl-workflows/REQ_cwl-workflows_adhoc-cwl-execution.adoc

```diff
@@ -0,0 +1,9 @@
+[requirement]
+====
+[%metadata]
+label:: /req/cwl-workflows/adhoc-cwl-execution
+description:: For executing a workflow defined using the Common Workflow Language (CWL) definition:
+part:: The Implementation SHALL support ad-hoc execution of a CWL workflow definition POSTed to `/processes/CWL/execution`
+====
+
+TODO: Verify whether a `{processId}` for the execution end-point is relevant or not for CWL.
```

Comment:
@fmigneault To clarify the TODO:
- this conformance class is specifically about ad-hoc execution of the CWL workflow **without** deploying it at all (either before, or as a result of submitting it).
- there is an assumption that it only references processes already defiend
- the thing to verify is whether it makes sense to POST to a particular `/processes/{IdOfProcessReceivingTheWorkflow}/execution` as opposed to a generic "CWL workflow execution end-point" as I currently specified (`/processes/CWL/execution`). My understanding is that there is no main process intended to receive a CWL workflow... the workflow itself defines the whole process and there is no "main" process in a CWL workflow. Would that be correct?

Reply:
In that case, a suggestion could be that the `id` for the CWL steps should be the URL of the existing processes locations. This would match the provided `"process": "<url>"` in proposed *Workflow* requirements.

You are correct. Neither `/processes/{IdOfProcessReceivingTheWorkflow}/execution` nor `/processes/CWL/execution` make sense. When CWL is used to define processes with Part 2, it is not a pre-existing process that receives the CWL to run it. The API itself parses the CWL, and generates a process each time. The engine running those processes can then execute them under the hood using any CWL-runner implementation from https://www.commonwl.org/implementations/, or even run them with a completely different approach if the CWL was converted to another equivalent definition.

Reply:
> nor /processes/CWL/execution make sense

Well, this conformance is about ad-hoc workflows specifically (because theoretically, deployed workflows are already handled with Part 2 and a dedicated conformance class for CWL). So we need some end-point at which to POST the CWL and respond right away with output, either as synchronous or asynchronous excution defined in Part 1, or as Collection Output defined in Part 3 (if that is supported). In order to be compatible with Part 1-Core and Part 3-CollectionOutput, a `/processes/{something}/execution` makes this easiest (at least for the extended Part 1 execution request approach).

We also discussed with Gerald something like `/workflows` instead which has some interesting aspects, but at least for the non-CWL/OpenEO approach breaks the idea that this simply extends the process execution request of _Part 1_ at `/proceses/{processId}/execution`.

> When CWL is used to define processes with Part 2, it is not a pre-existing process that receives the CWL to run it. The API itself parses the CWL, and generates a process each time. 

Yes because in that case you POST to `/processes`.
From the perspective of allowing nested processes to define a workflow by extending Part 1 execution requests, there is in that case always a top-level process which will generate the final outputs.

One option is to do ad-hoc workflow execution differently for ad-hoc workflows with OpenEO or CWL.
All things we could discuss on Friday 10:00 AM Eastern. Any chance you're available and interested to join?

Reply:
There might not be anything inherently wrong about the POST `/processes/CWL/execution` approach, but my impression is that if a server implements Part 2 with CWL understanding, it wants to offer additional capabilities that CWL does not support by itself (as detailed in the Best Practices document). For example, to provide additional I/O handling capabilities like parsing STAC metadata, FeatureCollections, dispatching steps to different servers, etc. In that case, such a server would purposely avoid using a generic single-app `CWL` process, since it would not offer the same extended capabilities to convert between OGC definitions and CWL definitions. Most implementations would tend to work in a mutually exclusive approach between deployed-CWL per-process vs ad-hoc CWL single-process. 

I also think that an ad-hoc process that takes an input CWL definition, to run it directly, is technically in no way different from a single-app Docker application. It just so happens that the `CommandLineTool` is a CWL-runner in that case. I wouldn't consider this process an actual *Workflow*, because the *OGC API - Processes* does not need to chain any step I/O by itself between distinct processes. It all runs under that single-app CWL-runner.

Therefore, `/processes/CWL/execution` could be a valid *recommendation*, but not necessarily a *requirement*.

I think it is preferable to not have a separate `/workflow` endpoint. The purpose of Part 2 when deploying single-apps or workflows is to provide a *Process* abstraction that can then be executed the same way with a common endpoint regardless of what the application package contains. This would defeat that purpose. I think Part 3 should also try to take advantage of this versatile abstraction.

Reply:
> It is in fact the opposite. If it supports Part 2 with CWL (i.e.: it has an engine that knows how to convert and run a CWL into individual and distinct processes), it does not need a generic process with another engine to run future CWL POSTed. It can simply reuse the already employed CWL-runner engine applied API-wide. However, if Part 2+CWL is not supported, then the single-app CWL Process makes sense (even if the API is only _Part 1: Core_ compliant without _Part 3_).

I am not sure what you mean by _single-app CWL Process_. For either Part 2+CWL or Part3+CWL, it doesn't matter how many processes or apps are involved. The difference between Part2+CWL and Part3+CWL is whether you first need to deploy the CWL to create a process in order to execute it (which may require authentication rights in order to create public resources on the server), vs. ad-hoc execution that does not require "write" rights to the server and could be offered to a wider range of users.

> Indeed. This is why I suggest the single-app CWL would not be required, just a suggestion. The way to handle new CWL would be left to the implementation. We cannot _require_ that the CWL single-app be the way to handle those new CWL.

Please explain what you mean by single-app. The CWL Part 1 vs. Part 2 approach is _"Deploy (for everyone), then Run"_, vs. "Just Run" (just for me to get the output). You can combine the two of course.

> It could, but the way I see it is that it would not be different from running multiple shell commands or Python functions. It is part of the Application Package abstraction that this _Process_ performs. On the _OGC API - Processes_ level, _Workflows_ should only consider the sequencing/parallelization of _Process_ entities themselves, not their underlying definitions.

Yes, I think this is true for Parts 1,2 and 3.

> I think that makes sense only if this Registry can run the Process (i.e.: it is itself an _OGC API - Processes_). Otherwise, the implementation receiving this URI reference to execute it would need to somehow know how to run all possible implementations of processes.

The idea of the WellKnown Processes Registry is to enable clients to discover processing resources able to accomplish a certain task (e.g., [rendering a map](https://github.com/opengeospatial/ogcapi-maps/issues/42), [calculating a route](https://github.com/opengeospatial/ogcapi-routes/blob/master/proposals/clause_8_ogcapi_processes.adoc), generating elevation contours, buffering vector geometry...) in a standardized manner (standardized inputs, outputs, and behvavior). Any discovered process that provides this capability should be suitable for the client. Such a registry could possibly to publicly available computing resources implementing those capabilities, but that would not be its primary purpose.

> That Registry would also end up duplicating most, if not all, _OGC API - Processes_ deployment schemas for validation of new _Process_ definitions, so I should just be one instead. I think `/processes` with advanced search capabilities makes a fine Registry.

The main thing this registry would do is provide you a _ProcessDescription_ (as you find at a `/processes/{processId}`) associated with a given URI. I am thinking of something e.g., on the OGC definition server like `http://www.opengis.net/def/WellKnownProcesses`. As we have for [TileMatrixSets](http://www.opengis.net/def/tms) and [link relation types](http://defs.opengis.net/vocprez/object?uri=http://www.opengis.net/def/rel).

The idea is to have processes with identical or equivalent processes offered by multiple implementations.

> 
> If the server _ONLY_ supports _Part 3_, then, indeed, that would be required to indicate where to POST the request (`/processes/CWL/execution`). That location should be a requirement. What I referred to as a recommendation is that conformance class itself that includes the `/processes/CWL/execution` requirement. If the server supports both _Part 2_ and _Part 3_, that "forced" requirement from _Part 3_ contradicts deployments from _Part 2_. This would also make it more confusing to the client about which strategy would actually be applied, since that would be an implementation detail.

It is definitely not the intent of that "CWL ad-hoc workflow execution" conformance class in Part 3 to add any requirement on how you deploy CWL workflows as a process using Part 2. Currently, the only thing the Part 3 CWL conformance class does is define ad-hoc execution of CWL ("Just Run" this for me only and give me the output). I suggest that a conformance class on how to deploy CWL workflows as a process be created in Part 2 -- but it could be in Part 3 as well, as a separate conformance class, it doesn't really matter where it is defined.

> 
> My intention is to support both _Part 2_ and _Part 3_. However, to run CWL components, I do not want 2 separate code paths as currently suggested (1 run by the API engine when deployed as _Part 2_, and 1 by the ad-hoc `CWL` process when executed by _Part 3_). For this reason, I would rather have _Part 3_ build CWL-related concerns on top of _Part 2_ definitions, rather than specify its own variant.

Well the whole point of the ad-hoc CWL conformance class in Part 3 is to "Just Run" without requirement prior deployment that Part 2 is all about. So I am not sure I understand your suggestion...

Reply:
By "single-app CWL", I refer to the reference ad-hoc `CWL` Process if `/processes/CWL/execute` is required by Part 3. One app that receives the CWL document (of any kind) and spawns a worker running (eg; `cwltool`) with it to accomplish the "Just Run". Technically, the CWL contents could be a single `class: CommandLineTool` or a `class: Workflow` (in CWL context, not * Part 3* OGC context), so it can exist both on *Part 1*-only or *Part 1+3* servers.  

My main point is that if the server supports Part 2 Deploy+CWL, then it inherently works with variable  `/processes/{processId}/execute` creation with submitted CWL (whether in "Jun Run" or "Deploy+Run" fashion). Therefore, if Part 3 _**requires**_ that always the same Process on `/processes/CWL/execute` is used, they work in contradicting manner.

A server supporting only Part 1 could have an all-purpose single-app `CWL` process. However, that `CWL` process loses all its meaning if the server itself understands CWL via Part 2. The *Part 2* server doesn't need to dispatch the execution to an intermediate `CWL` process that can run a Workflow, it can already do it itself. 

All those points also apply in the case of OpenEO when it comes to handling Workflow definitions. 

For *Well-Known Processes Registry*, the problem I see with that approach is when it comes to running the real Process, the server needs its concrete implementation. Calculations over a matrix of pixels like NDVI could be "simple enough" to be described with some text representation in the Registry, but anything more (e.g.: an ML training pipeline) becomes impossible to describe in a replicable manner, unless some Docker or similar is involved. However, there is not a "common" way to do that (I wouldn't trust another `ml-trainer` to implement the correct training I need for my use case), so publishing to a *Well-Known Processes Registry* seems like a misuse of reference definitions. I also doubt the registry would be updated every time a user develops a new app. Therefore, those apps in development that are not published to the Registry could not make use of the `"process": "<ref>"` format if Registry use is enforced. 

Using a `"process": "<any-url>"` approach would resolve the above problems. 
If, like you mentioned, `registry would do is provide you a ProcessDescription`, then I see no reason not to use an *OGC API - Processes* server instance directly, instead of going through this intermediate Registry entity.

Reply:
> My main point is that if the server supports Part 2 Deploy+CWL, then it inherently works with variable /processes/{processId}/execute creation with submitted CWL (whether in "Jun Run" or "Deploy+Run" fashion).

Do you mean by this that with Part 2+CWL, you have the ability to deploy some CWL process that then allows you to POST/deploy a process at some `{processId}` that will then be able to execute arbitrary CWL to execute ad-hoc workflows? If so, the use case that this does not cover, that Part 3+CWL is about, is for users that are not authorized to deploy new processes, and there is no pre-existing CWL runner process, but whom you would still like to allow to "Just Run" CWL workflows. If that is not what you meant then probably we should try to discuss this live... There are a lot of flexibility from all these Processes parts and CWL, and it is difficult to read each other's mind in GitHub issues :)

> (e.g.: an ML training pipeline) becomes impossible to describe in a replicable manner

Depending on the nature and complexity of the WellKnownProcess, different implementations may replicate the results to a different degree of similarity/equivalence. Part of the WellKnownProcess description could document the reproducibility aspects. We are looking at the reproducibility aspects specifically in _Testbed 18 - Identifiers for Reproducible Science_, which has a lot to do about having specific DOI for Dockers etc.

I brought up WellKnownProcesses in this discussion here in the context of a "CWL Runner" process, which would have a fairly standard behavior based on the CWL specification. Other things like NDVI, contour generation should also be fairly reproducible. Calculating a route and rendering a map may produce different but similar / equivalent results. The reproducibility aspects are important, but not the primary objective of this registry -- rather it is the discovery of a particulary capability.

We are not looking for a specific process implementation, but to recognize different implementations performing the same type of calculation from the same type of inputs, which should produce at least similar results. 

>  I also doubt the registry would be updated every time a user develops a new app.

It is not so much intended for every possible app being developed, but rather for the most common type of calculations for which there are several different implementations, for which there might be several different deployments. Also for the kind of smaller processing steps that could be organized together in a workflow using Part 3 ad-hoc workflows.

It may or may not be applicable to machine learning processing steps.

I think there would be benefit to a dedicated extension to Processes for training ML Models.
Several of the conformance classes of Part 3 makes sense for preparing the inputs to the trainer (_Collection Input/Remote collections, Nested/Remote Processes, Input/Output Fields modifiers_), as well as the expected values. But we are missing a first class ML model object (that could possibly have a corresponding API resource) that gets progressively refined as you train it with more and more data. The Part 1 Process output is not a good fit for this, so really the only practical solution is something behind the scenes that is not standardized, and that the client has no or limited visibility into.

I think one could create an ML model with an API that references a Part 1 process that could be used to performs the actual training, calculating the predicted vs. actual differences and doing the back-propagation (which perhaps could even be a well-defined WellKnownProcess, using a particular algorithm), but could also leverage workflow definitions in order to specify the inputs and expected outputs, and any pre-processing to apply. Then you could have something similar to _Collection Output_ where you can train for a specific area/time/resolution of interest, but that would not actually return any output (except perhaps current stats for the model), but result in a refinement of the model. The ML model API could also offer functionality to refer to the model from an associated prediction process, could offer CRUD to import/update/export/delete models, maybe even conformance classes for versioning and comparing predictions from different versions of the model.

There is a [thread in Testbed 18  looking specifically at Machine Learning training](https://portal.ogc.org/files/?artifact_id=100034#ML), but I have not been able to follow what is happening there.

However, I think the current _Collection Output_ and on-demand triggering of Part 3 is well suited to the prediction process once the model is trained.

> Therefore, those apps in development that are not published to the Registry could not make use of the "process": "<ref>" format if Registry use is enforced.

It is certainly not the intent that every single process needs to be registered as a WellKnownProcess.
But implementations that provide processes that do conform to a particular WellKnownProcess could advertise that they do in those processes descriptions with a `"wellKnownProcessURI" : "http://www.opengis.net/def/wkprocesses/OGC/1.0/CWLRunner"` to identify as a CWL Runner, or `"wellKnownProcessURI" : "http://www.opengis.net/def/wkprocesses/OGC/1.0/RenderMap"` to identify as a map rendering process, or `"wellKnownProcessURI" : "http://www.opengis.net/def/wkprocesses/OGC/1.0/CalculateRoute"` to identify as a route calculator, or `"wellKnownProcessURI" : "http://www.opengis.net/def/wkprocesses/OGC/1.0/NDVI"` to identify as an NDVI calculator (with an expectation of a particular process description and similar output, regardless of the implementation / deployment).

Reply:
> Do you mean by this that with Part 2+CWL, you have the ability to deploy some CWL process that then allows you to POST/deploy a process at some `{processId}` that will then be able to execute arbitrary CWL to execute ad-hoc workflows?

Technically, yes, we could deploy a "meta-CWL" process whose job is to run other CWL documents. However, those other CWL documents would not create a new `{processId}` unless the "meta-CWL" explicitly does a POST request to the server's `/process` endpoint. Also, the CWL documents run by this "meta-CWL" process would not profit from extended capabilities offered by the OGC-API server, since it would run with a "native" `cwltool`. 

> [...] for users that are not authorized to deploy new processes [...]

My opinion about this is that, if the server has a good enough reason to block users from deploying new processes with CWL, I don't see why it would allow them to "bypass" deployment and achieving a similar result with immediate execution. It seems like a security design issue. CWL have to be "deployed" (minimally saved to disk or loaded in the engine) before running them. We cannot do "Just Run" by plugging the values directly.
It can make sense in a *Part 1+3* server to allow the "meta-CWL" to resolve the "Just Run" issue, but seems odd in a *Part 1+2+3* server.

About *WellKnownProcess*, it is a valid use case (I wasn't saying otherwise). I was just explicitly stating that it should not be the *exclusive* method to reference processes. So, we should allow `"process": "<ref-WellKnownProcess|url-DescribeProcess>"`. If that is permitted, then I have no problem with it. Since additional requests handling steps would be needed to query the Registry and obtain the *ProcessDescription* from it (rather than directly accessing a URL to a *ProcessDescription*), maybe a separate requirement could be defined for servers to indicate if they support Registry handling? 

> dedicated extension to Processes for training ML Models [...]

Yeah. If recommendations come up at some point, it could be a good addition. There are full working groups dedicated to this, as well as many other Workflow initiatives like KubeFlow, but it is sadly far from getting close to a federated and replicable solution. There are options, but it is mostly custom definitions for each experiment.

We could indeed discuss it live to explore all avenues.

Reply:
> It can make sense in a Part 1+3 server to allow the "meta-CWL" to resolve the "Just Run" issue, but seems odd in a Part 1+2+3 server.

Part 3+CWL is all about _Just Run_ and this meta-CWL capability (whether there is an actual meta-CWL process implemented using Part 1, or the functionality to execute CWL is just associated with an arbitrary end-point). The use case for a Part 2 server is just to allow those (but perhaps I make some wrong assumptions about a non-existing of CWL to execute local and/or remote OGC API - Processes, regardless of their implementation).

> It seems like a security design issue. CWL have to be "deployed" (minimally saved to disk or loaded in the engine) before running them. 

If there is some kind of sandboxing environment, a controlled limit on resources, it might be safe enough to offer this capability while not wanting users to pollute a pre-defined set `/processes`. I might also be making wrong assumption about CWL's capabilities as a workflow language that could be restricted to only run a existing processes and remote processes from pre-authorized sources, or even a restricted set of approved docker images, as opposed to running arbitrary code (which has much greater safety concerns).

> "process": "<ref-WellKnownProcess|url-DescribeProcess>"

This is an interesting misunderstanding. In a workflow definition, the intent was never (so far) to reference a WellKnownProcess from a workflow.

The purpose of the WellKnownProcess is to include it in a **_ProcessDescription_**, as `"wellKnownProcessURI" : "..."`.

But here you make a point that perhaps for defining a re-usable workflow you don't care about a specific implementation being used, it does make sense to specify a WellKnownProcess as an alternative (assuming the workflow execution as some mechanism to do the discovery itself, such as the registry having a list of good and reliable implementations of it).

Reply:
> assumptions about [...] CWL to execute local and/or remote OGC API - Processes

That's the thing. CWL by itself cannot run *OGC API - Processes* requests. You need a literal application definition (docker, python script, etc.) that sends HTTP requests to execute parsed inputs, poll status, and retrieve results. That application can then be used to chain process in a Workflow. The server must perform the conversion of the submitted CWL definition into one that can use this "*OGC API - Processes* requests handler" with the processes generated from the original CWL definition. This is part of the "additional capabilities" I was referring to. There are other extra capabilities, such as handling special file-types Collection I/O or STAC I/O into a `Directory` that would be needed with Part 3.

Reply:
Thanks for clarifying. What I would imagine is that a CWL workflow definition could be extended or some existing URI could be re-used to refer to _OGC API - Processes_, local or remote.

Then the OGC API - Processes implementation receiving the CWL workflow (whether it's deployed with Part 2, or executed ad-hoc with Part 3), could automatically translate it to this more explicit version that a traditional CWL runner is able to execute (or just understand this natively, if it is its own implementation of CWL).

Such extensions allowing to refer to _OGC API - Processes_ URIs would be in scope for the CWL conformance class of Part 3, and could be developed in collaboration with the CWL team so that it could also be an official extension of CWL itself at some point. Same for OpenEO.

A similar approach for providing a more streamlined access to OGC API Collections and STAC could also work.

I think this restriction on workflows allowing to run pre-defined or pre-authorized processes, and built-in collection input/output, still allows to build complex powerful workflows but without providing the ability to explicitly and directly run executable code or write to the disk, and therefore could be somewhat safer than a workflow that can execute arbitrary code or executables from anywhere.

That is an important part of the rationale behind the Part 3 ad-hoc workflows idea. (good topics to add to the still empty _Security Considerations_ section!)